sympy/physics/quantum/lapply.py

```diff
@@ -0,0 +1,1434 @@
+        cr, ncr, ar = lapply_pow_roll1_c_nc_ncef(rhs.base, rhs.exp,
+                to_the_left, force=True, cached=cached, **options)
+        lhsar = lapply_mul2([lhs, ar], [], **options)
+        if not lapply_no_act(lhs, ar, lhsar):
+            # interaction; return resulting factors as list..
+            return cr + [lhsar] + ncr   # ..instead of Mul()
+    return None
+
+# Create (empty) Dispatcher for layer B of rules:
+# Layer B contains handlers that either have lower priority
+# than layer A, or would create a type ambiguity in Dispatcher.
+lapply_mul2types_B = Dispatcher("lapply_mul2types_B")
+# See sympy.physics.quantum.qapply for examples for layer B.
+
+
+# -----------------------------------------------------------------------------
+# Cache for Pow
+# -----------------------------------------------------------------------------
+# The purpose of caching is to avoid frequent recomputation if a Pow is
+# rolled off to Pow(base,exp-1)*base -> Pow(base,exp-2)*base*base etc.
+#
+# Principles of Caching for base and exponents of Powers:
+# - Only cache exprs that will probably be used multiple times. Do not litter the
+#   the cache with one-time results. So bases of Powers are cached. Exponents are
+#   cached only in case of option["apply_exp"] and non-simple exponents.
+# - Exprs being cached must have been returned in ncl from lapply_mul2_c_nc, so
+#   all its factors have been processed fully.
+# - The cache returns three types of information on an expr (besides None):
+#   i.   "elem": expr is elementary  ii. "expr": expr is fully processed
+#   iii. (rolloff, cl, encl) expr is base of a power.
+#   which roughly correspond to output of c_nc_ncef, lapply_mul2, lapply_mul2_c_nc.
+# - Base and exponents of Pows are cached separately. If a power is rolled off
+#   the exponents counted down are removed from cache to avoid litter. This may
+#   lead to re-computation of exponents in situations as Pow(A,exp)*B+Pow(A,exp)*C
+#   which is unavoidable as the cache only has one instance.
+
+class RO(Enum): # the roll-off type of a base:
+    UNKN = 0    # Roll-off type is unknown
+    NORO = 1    # base*base doesn't interact, so don't roll off base**exp
+    ROLL = 2    # base*base interacts, so roll off base**exp off
+
+# Pow_with_Cache returns Pow(base, exp), but also caches the rolloff
+# decision assigned to base (rolloff) and the splitting of base
+# into cl and elementary ncl factors (as derived from lapply_mul2_c_nc).
+def Pow_with_Cache(base:Expr, exp:Union[Expr, None], rolloff:RO,
+                   cl:List, encl:List, **options) -> Union[Expr, Tuple]:
+    """Caches factorization data for base. If exp is None, returns the
+    cached tuple (rolloff, cl, encl) else Pow(base, exp)."""
+
+    if exp is not None:  # exp==None is to register a base only
+        if   exp is S.Zero: p = S.One
+        elif exp is S.One : p = base
+        else: p = Pow(base, exp, evaluate=False)

```

Comment:
Again here things are on one line when it would be easier to read in the standard Python style:
```python
    if exp is not None:  # exp==None is to register a base only
        if exp is S.Zero:
            p = S.One
        elif exp is S.One:
            p = base
        else:
            p = Pow(base, exp, evaluate=False)
```
We don't need to worry about adding spaces to align everything because the natural indentation structure is supposed to create the alignment that clearly shows that e.g. each branch assigns to `p`.